nb2kg/handlers.py

```diff
@@ -222,8 +229,15 @@ def _read_messages(self, callback):
             else:  # ws cancelled - stop reading
                 break
 
-        if not self.disconnected: # NOTE(esevan): if websocket is not disconnected by client, try to reconnect.
-            self.log.info("Attempting to re-establish the connection to Gateway: {}".format(self.kernel_id))
+        # NOTE(esevan): if websocket is not disconnected by client, try to reconnect.
+        if not self.disconnected and self.retry < KG_WS_RETRY_MAX:
+            # exponential backoff to retry
+            jitter = random.randint(10, 100) * 0.01
+            retry_interval = KG_WS_RETRY_INTERVAL * (2 ** self.retry) + jitter
+            self.retry += 1
+            self.log.info("Attempting to re-establish the connection to Gateway in %s secs (%s/%s): %s",
+                          retry_interval, self.retry, KG_WS_RETRY_MAX, self.kernel_id)
+            yield gen.sleep(retry_interval)
```

Comment:
This looks fine.  My only comment, and assuming my math is correct, the 10 retries will take just over 17 minutes to complete with 8.5 minutes between the last two retries.  I think if we lowered the max retries to 6, the sequence will be completed in just over a minute - with 30 seconds between the last two retries.  This seems sufficient to me - especially in these kinds of situations (where recovery is seldom successful). 

If you feel more time is warranted, I'd be okay going one, maybe two, more iteration(s).

Reply:
Oops.. I've never calculated this.. Actually I used only 3 for default in my env. I'll add 30 seconds max interval and reduce the max retries to 5(I like prime number :grin:) if it is Okay.

I used exponential backoff because k8s recovered EG quite randomly; sometimes 3s or sometimes a couple of minutes.